### ISSUE TITLE
Multiple access to keychain in a single session

### ISSUE THREAD

**Issue by deepak1556** [microsoft/vscode#107480]
When testing https://github.com/microsoft/vscode/pull/107292

* Delete `vscode-insiders-github.login` from keychain
* Download insiders which has keytar in the renderer, for this test I used  `f943318042` https://az764295.vo.msecnd.net/insider/f943318042a3b9229049aea86cea04d0ff9c8210/VSCode-darwin-insider.zip
* Install github PR extension and login to github
* Check `vscode-insiders-github.login` entry is created in the keychain
* Update and restart the insiders which will have keytar in the main process
* New password prompt to access keychain
* Enter password and click `Allow`
* Every 30s the prompt appears as long as `Allow` option is used
* `Always Allow` does stop the repeated prompts

I would expect the keychain is accessed only once per session for the PR extension ? Is this intended behavior ?

**Comment by RMacfarlane** [comment]
Currently, yes. 

Since there currently is no event to let an extension know that the keychain has been changed, each window polls to see if a user has signed in or out from another window.

After adding `getPassword` and `setPassword` to vscode's API, we can also introduce a change event so that we no longer have to do this

**Comment by deepak1556** [comment]
> Since there currently is no event to let an extension know that the keychain has been changed, each window polls to see if a user has signed in or out from another window.

Got it, thanks! Now with a single source of access to the keychain (main process) the change event would be really good to have. Is there a separate issue tracking this work or do you want to use this one ?

**Comment by RMacfarlane** [comment]
Let's keep this open and use it to track that

**Comment by bpasero** [comment]
If there is something I should do here please assign me back, but I just moved `keytar` to the main process and otherwise do not feel like an owner.

**Comment by rebornix** [comment]
What happened to me today:

* Install Insiders update
* An authentication prompt, I choose "always allow"
* Open extensions page, GHPRI has an update, I click reload the window
* Another prompt shows up

**Comment by deepak1556** [comment]
Consider @RMacfarlane is on vacation this week and the issue can't be addressed for current milestone, @bpasero can we revert the keytar change for this milestone, otherwise the experience with repeated keychain access can annoy users easily.

**Comment by bpasero** [comment]
@deepak1556 do you have any idea what is causing this? I tried the steps with a window reload and cannot reproduce.

@rebornix anything special with your setup?

**Comment by bpasero** [comment]
@rebornix please check how your keychain is configured:

![image](https://user-images.githubusercontent.com/900690/94534180-5f99a780-0240-11eb-98ee-d64b0787ec2f.png)

**Comment by rebornix** [comment]
@bpasero 

<img width="646" alt="image" src="https://user-images.githubusercontent.com/876920/94604285-4bf43e00-024c-11eb-9e96-284f0fec1688.png">

This is identical to all other keychain items I have and I think they are default values. @deepak1556 ping me if you can't reproduce as I'm seeing it every time I get a new Insiders update. Also I saw a second prompt today when switching workspaces.

**Comment by chrmarti** [comment]
Could this be 1.49 accessing via renderer and 1.50-insiders accessing via main process stepping on each others toes?

**Comment by bpasero** [comment]
@chrmarti maybe, @rebornix did you possibly also run stable and get prompted?

**Comment by rebornix** [comment]
@bpasero I'm not using Stable. I no longer see this issue today, suspicion is something is wrong with my macOS Catalina but I don't have a proof. I'll say it's not a blocker as I'm probably the only one who is seeing this issue constantly everyday before today.

**Comment by bpasero** [comment]
Ok sounds good, closing thereby.

**Comment by bpasero** [comment]
Reopening for the original issue Deepak reported.

**Comment by SachinShekhar** [comment]
@RMacfarlane 

> After adding `getPassword` and `setPassword` to vscode's API, we can also introduce a change event so that we no longer have to do this.

Why did you guys push this half-baked change to stable channel?

This is inconvenience to multiple users. See https://github.com/microsoft/vscode/issues/108342 https://github.com/microsoft/vscode/issues/108352

**Comment by Blaisorblade** [comment]
> Why did you guys push this half-baked change to stable channel?

Here was the reasoning I guess (emphasis mine):
https://github.com/microsoft/vscode/pull/107292#pullrequestreview-494963387

> This makes sense, I think it's good that the password prompt is much clearer now and _not too disruptive since it should only be shown once_

**Comment by SachinShekhar** [comment]
@Blaisorblade This issue was opened after that comment. This issue clearly says that password prompt is being shown every 30 seconds.

Without fixing this issue, why would anyone push this to stable channel?

**Comment by dboyce-gh** [comment]
I am seeing this on stable 1.50 too.  Makes Code unusable on Mac unless you 'allow all'.
